### ISSUE TITLE
Fatal error occurs when calling set_current_gateway() and there is no WC session available

### ISSUE THREAD

**Issue by hjoelr** [woocommerce/woocommerce#25241]
**Describe the bug**
Calling `WC()->payment_gateways()->set_current_gateway( $gateways )` when there is no session WC session object causes a fatal error to occur on the following line:

https://github.com/woocommerce/woocommerce/blob/master/includes/class-wc-payment-gateways.php#L186

**To Reproduce**
1. Add the following code to functions.php.

```php
if ( ! wp_doing_ajax() ) {
	add_action( 'admin_init', 'test_set_current_gateway', 11 );
	function test_set_current_gateway() {
		$payment_gateways = WC()->payment_gateways()->payment_gateways();
		WC()->payment_gateways()->set_current_gateway( $payment_gateways );
	}
}
```

2. Navigate and login to the WC backend.
3. Observe WC crash because `set_current_gateway()` is calling `get()` on a null session object.

**Expected behavior**
No crash is experienced.

**Isolating the problem (mark completed items with an [x]):**
- [ ] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [ ] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

**WordPress Environment**
<details>
```

### WordPress Environment ###

WordPress address (URL): https://store.local.dev/store
Site address (URL): https://store.local.dev/store
WC Version: 3.8.1
REST API Version: ✔ 1.0.2
Log Directory Writable: ✔
WP Version: 5.3.1
WP Multisite: –
WP Memory Limit: 128 MB
WP Debug Mode: ✔
WP Cron: –
Language: en_US
External object cache: ✔

### Server Environment ###

Server Info: Apache/2.4.38 (Debian)
PHP Version: 7.2.25
PHP Post Max Size: 96 MB
PHP Time Limit: 0
PHP Max Input Vars: 1000
cURL Version: 7.64.0
OpenSSL/1.1.1d

SUHOSIN Installed: –
MySQL Version: 5.7.28
Max Upload Size: 96 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ❌ Your server does not have the SoapClient class enabled - some gateway plugins which use SOAP may not work as expected.
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 3.8.1
WC Database Prefix: ce0p3_
Total Database Size: 666.59MB
Database Data Size: 476.15MB
Database Index Size: 190.44MB
ce0p3_woocommerce_sessions: Data: 2.52MB + Index: 0.08MB + Engine InnoDB
ce0p3_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
ce0p3_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
ce0p3_woocommerce_downloadable_product_permissions: Data: 0.52MB + Index: 1.00MB + Engine InnoDB
ce0p3_woocommerce_order_items: Data: 2.52MB + Index: 1.52MB + Engine InnoDB
ce0p3_woocommerce_order_itemmeta: Data: 23.56MB + Index: 23.09MB + Engine InnoDB
ce0p3_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
ce0p3_woocommerce_tax_rate_locations: Data: 0.08MB + Index: 0.14MB + Engine InnoDB
ce0p3_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
ce0p3_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
ce0p3_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
ce0p3_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
ce0p3_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
ce0p3_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
ce0p3_akeeba_common: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
ce0p3_ak_params: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
ce0p3_ak_profiles: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
ce0p3_ak_stats: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
ce0p3_ak_storage: Data: 11.52MB + Index: 0.00MB + Engine InnoDB
ce0p3_ak_users: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
ce0p3_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
ce0p3_comments: Data: 19.55MB + Index: 17.09MB + Engine InnoDB
ce0p3_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
ce0p3_options: Data: 1.52MB + Index: 0.17MB + Engine InnoDB
ce0p3_pmxe_exports: Data: 0.33MB + Index: 0.00MB + Engine InnoDB
ce0p3_pmxe_google_cats: Data: 0.39MB + Index: 0.00MB + Engine InnoDB
ce0p3_pmxe_posts: Data: 1.52MB + Index: 0.00MB + Engine InnoDB
ce0p3_pmxe_templates: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
ce0p3_pmxi_files: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
ce0p3_pmxi_history: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
ce0p3_pmxi_imports: Data: 1.52MB + Index: 0.00MB + Engine InnoDB
ce0p3_pmxi_posts: Data: 1.52MB + Index: 0.00MB + Engine InnoDB
ce0p3_pmxi_templates: Data: 0.09MB + Index: 0.00MB + Engine InnoDB
ce0p3_postmeta: Data: 102.64MB + Index: 101.36MB + Engine InnoDB
ce0p3_posts: Data: 272.78MB + Index: 14.42MB + Engine InnoDB
ce0p3_relevanssi: Data: 2.52MB + Index: 6.06MB + Engine InnoDB
ce0p3_relevanssi_log: Data: 4.52MB + Index: 0.00MB + Engine InnoDB
ce0p3_relevanssi_stopwords: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
ce0p3_termmeta: Data: 0.11MB + Index: 0.17MB + Engine InnoDB
ce0p3_terms: Data: 0.11MB + Index: 0.16MB + Engine InnoDB
ce0p3_term_relationships: Data: 4.52MB + Index: 3.52MB + Engine InnoDB
ce0p3_term_taxonomy: Data: 0.11MB + Index: 0.14MB + Engine InnoDB
ce0p3_usermeta: Data: 18.55MB + Index: 19.06MB + Engine InnoDB
ce0p3_users: Data: 1.52MB + Index: 0.70MB + Engine InnoDB
ce0p3_vtprd_lifetime_limits_purchaser: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
ce0p3_vtprd_lifetime_limits_purchaser_logid_rule: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
ce0p3_vtprd_lifetime_limits_purchaser_rule: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
ce0p3_vtprd_purchase_log: Data: 0.23MB + Index: 0.02MB + Engine InnoDB
ce0p3_vtprd_purchase_log_product: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
ce0p3_vtprd_purchase_log_product_rule: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
ce0p3_wc_download_log: Data: 0.16MB + Index: 0.14MB + Engine InnoDB
ce0p3_wc_product_meta_lookup: Data: 0.38MB + Index: 0.81MB + Engine InnoDB
ce0p3_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
ce0p3_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
ce0p3_woocommerce_bundled_itemmeta: Data: 0.23MB + Index: 0.31MB + Engine InnoDB
ce0p3_woocommerce_bundled_items: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
ce0p3_wpfront_ure_login_redirect: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
ce0p3_wpfront_ure_options: Data: 0.03MB + Index: 0.00MB + Engine InnoDB

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (47) ###

User Switching in Admin Bar: by Mark Wilkinson – 1.1.1
Akeeba Backup for WordPress: by Akeeba Ltd – 3.6.1
Better Search Replace: by Delicious Brains – 1.3.3
Black Studio TinyMCE Widget: by Black Studio – 2.6.9
Classic Editor: by WordPress Contributors – 1.5
Coming Soon Page, Under Construction & Maintenance Mode by SeedProd: by SeedProd – 5.1.0
Cue: by AudioTheme – 2.3.3
Disable REST API: by Dave McHale – 1.5
Enable Media Replace: by ShortPixel – 3.3.7
Enhanced Media Library: by wpUXsolutions – 2.7.2
Homepage Control: by WooThemes – 2.0.3
Mailgun: by Mailgun – 1.7.1
MCE Table Buttons: by Jake Goldman 10up Oomph – 3.3
Regenerate Thumbnails: by Alex Mills (Viper007Bond) – 3.1.2
Relevanssi: by Mikko Saari – 4.3.4
Fast Secure Contact Form: by fastsecure – 4.0.56
Simple Image Widget: by Cedaro – 4.4.2
Storefront Site Logo: by Wooassist – 1.2.3
TW Dev Store Modifications: by The Wilds – 1.1.1 – Not tested with the active version of WooCommerce
WooCommerce - Sage SData Integration: by Andy Anglea – 1.4.6
User Switching: by John Blackbourn & contributors – 1.5.3
Widget Context: by Kaspars Dambis – 1.2.0
WooCommerce Amazon S3 Storage: by WooCommerce – 2.1.15
WooCommerce Print Invoice & Delivery Note: by Tyche Softwares – 4.5.3 – Not tested with the active version of WooCommerce
WooCommerce - Store Exporter: by Visser Labs – 2.3.1
WooCommerce Authorize.Net Gateway: by SkyVerge – 3.1.1 – 3.1.2 is available
WooCommerce Purchase Order Payment Gateway: by WooCommerce – 1.2.5
WooCommerce Google Analytics Pro: by SkyVerge – 1.8.6
WooCommerce Pre-Orders: by WooCommerce – 1.5.20
WooCommerce Product Bundles: by SomewhereWarm – 5.14.0 – 5.14.2 is available
WooCommerce Protected Categories: by Barn2 Media – 2.3.1
WooCommerce Shipment Tracking: by WooCommerce – 1.6.17
WooCommerce UPS Shipping: by WooCommerce – 3.2.18
WooCommerce USPS Shipping: by WooCommerce – 4.4.35
WooCommerce Smart Coupons: by StoreApps – 4.5.0
WooCommerce Social Login: by SkyVerge – 2.8.3 – 2.8.4 is available
WooCommerce Wholesale Order Form: by Rymera Web Co – 1.14.3
WooCommerce Wholesale Prices Premium: by Rymera Web Co – 1.23.8
WooCommerce Wholesale Prices: by Rymera Web Co – 1.10 – Not tested with the active version of WooCommerce
WooCommerce WishLists: by Lucas Stark – 2.1.14 – 2.1.15 is available – Not tested with the active version of WooCommerce
WooCommerce: by Automattic – 3.8.1
WP Crontrol: by John Blackbourn & crontributors – 1.7.1
WP Last Login: by Konstantin Obenland – 3
WPFront User Role Editor: by Syam Mohan – 2.14.3

### Inactive Plugins (18) ###

### Dropin Plugins (1) ###

object-cache.php: object-cache.php

### Settings ###

API Enabled: ✔
Force SSL: ✔
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: bundle (bundle)
external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: ✔

### WC Pages ###

Shop base: #181 - /
Cart: #191 - /cart/
Checkout: #701 - /checkout/
My account: #711 - /my-account/
Terms and conditions: #8543 - /terms-conditions/

### Theme ###

Name: Online Store
Version: 1.3.0
Author URL: 
Child Theme: ✔
Parent Theme Name: Storefront
Parent Theme Version: 2.5.3
Parent Theme Author URL: https://woocommerce.com/
WooCommerce Support: ✔

### Templates ###

Overrides: online-store/woocommerce/cart/cart-shipping-checkout.php
online-store/woocommerce/components/cart-totals/cart-totals.php
online-store/woocommerce/cart/cart.php
online-store/woocommerce/cart/mini-cart.php
online-store/woocommerce/checkout/form-coupon.php
online-store/woocommerce/checkout/form-login.php
online-store/woocommerce/checkout/payment-method.php
online-store/woocommerce/checkout/review-order.php
online-store/woocommerce/emails/customer-new-account.php
online-store/woocommerce/emails/customer-on-hold-order.php
online-store/woocommerce/emails/customer-reset-password.php
online-store/woocommerce/emails/email-customer-details.php
online-store/woocommerce/emails/plain/customer-on-hold-order.php
online-store/woocommerce/emails/plain/customer-reset-password.php
online-store/woocommerce/global/quantity-input.php
online-store/woocommerce/loop/add-to-cart.php
online-store/woocommerce/loop/orderby.php
online-store/woocommerce/myaccount/dashboard.php
online-store/woocommerce/single-product/add-to-cart/grouped.php
online-store/woocommerce/single-product/title.php


### Product Bundles ###

Database Version: 5.14.0
Template Overrides: online-store/woocommerce/single-product/bundled-item-image.php

### Authorize.Net Credit Card ###

Environment: Test
Tokenization Enabled: ✔
Debug Mode: Off
Webhooks Enabled: ❌
						No signature key configured

### Social Login ###

Facebook: ✔ Available
Twitter: Disabled
Google: ✔ Available
Amazon: Disabled
Linkedin: Disabled
Paypal: Disabled
Instagram: Disabled
Disqus: Disabled
Yahoo: Disabled
Vkontakte: Disabled

### Action Scheduler ###

Complete: 4
Oldest: 2019-11-06 11:01:02 -0500
Newest: 2019-12-04 15:48:04 -0500

Pending: 0
Oldest: –
Newest: –

Canceled: 0
Oldest: –
Newest: –

In-progress: 0
Oldest: –
Newest: –

Failed: 0
Oldest: –
Newest: –

```
</details>

**Comment by hjoelr** [comment]
Let me explain what spurred this issue and PR. I'm using the WooCommerce Smart Coupons: by StoreApps plugin on my store. It adds some ability per coupon to limit coupons based on payment method. It calls `WC()->payment_gateways->get_available_payment_gateways()` on the coupon management screen to list available payment methods by which we could limit the coupons. This in turn calls the `woocommerce_available_payment_gateways` filter. I have another extension that filters payment gateways by role and is hooked on that filter and eventually calls `WC()->payment_gateways()->set_current_gateway( $gateways )`. Since this is all while logged into the admin interface, there isn't a `WC()->session` object.

Now, I don't even know if it makes sense for `set_current_gateway( $gateways )` to be called from the backend of WC. If not, then this might be a bug in the role filtering extension I'm using. However, if it's not improper for this function to be called from the backend, then it's a bug in WC. If anyone knows the proper usage of this function, I'd be glad to know and get with the extension developer if that's where the bug really lies.

**Comment by hjoelr** [comment]
I've updated the issue description with clearer instructions on how to replicate the error.

**Comment by juliaamosova** [comment]
Hi @hjoelr,

Thank you for reporting the issue. I can reproduce it on my end. When I add the snippet you provided to `functions.php`, I am getting the following:

![error](https://user-images.githubusercontent.com/19143190/71409313-03628100-2639-11ea-9e40-a0f53c22c96c.jpg)

My error log contains the following:

```
An error of type E_ERROR was caused in line 186 of the file /srv/htdocs/wp-content/plugins/woocommerce/includes/class-wc-payment-gateways.php. Error message: Uncaught Error: Call to a member function get() on null in /srv/htdocs/wp-content/plugins/woocommerce/includes/class-wc-payment-gateways.php:186
Stack trace:
#0 /srv/htdocs/wp-content/plugins/code-snippets/php/snippet-ops.php(361) : eval()'d code(5): WC_Payment_Gateways->set_current_gateway(Array)
#1 /wordpress/core/5.3.2/wp-includes/class-wp-hook.php(288): test_set_current_gateway('')
#2 /wordpress/core/5.3.2/wp-includes/class-wp-hook.php(312): WP_Hook->apply_filters(NULL, Array)
#3 /wordpress/core/5.3.2/wp-includes/plugin.php(478): WP_Hook->do_action(Array)
#4 /wordpress/core/5.3.2/wp-admin/admin.php(170): do_action('admin_init')
#5 {main}
  thrown
```

I am not immediately sure where the bug is coming from. However, let me mark your PR for review so that the PR reviewer could take a look further. Thank you for your contribution.

**Comment by rodrigoprimo** [comment]
Fixed in #25322
